### DISCUSSION TITLE
Extend Argument List for `ekg-note-create`

### DISCUSSION THREAD

**Discussion by qingshuizheng** [ahyatt/ekg#74] [Ideas]
Hi @ahyatt ,

Could we extend the argument list of `ekg-note-create`? After the following changes we could avoid some `setf` job if we decide to add properties or change subject thereafter (like what happens in `ekg-capture`).

For better diff reading, I've pushed the commit here: https://github.com/qingshuizheng/ekg/commit/941f27a30d

1. `ekg-note-create`: 

```elisp
;; Before:
(defun ekg-note-create (text mode tags)
  "Create a new `ekg-note' with TEXT, MODE and TAGS."
  (let* ((time (time-convert (current-time) 'integer)))
    (make-ekg-note :id (ekg--generate-id)
                   :text text
                   :mode mode
                   :tags tags
                   :creation-time time
                   :modified-time time)))

;; After:
(defun ekg-note-create (text mode tags &optional properties subject)   ;; <-- arg list is extended
  "Create a new `ekg-note' with TEXT, MODE, TAGS, INLINES, PROPERTIES and SUBJECT."
  (let* ((time (time-convert (current-time) 'integer)))
    (make-ekg-note :id (or subject (ekg--generate-id))
                   :text text
                   :mode (or mode ekg-capture-default-mode)            ;; <-- if mode is nil, set default
                   :tags tags
                   :properties properties                              ;; <-- properties is added
                   :creation-time time
                   :modified-time time)))

```

2. `ekg-capture` :

```elisp
;; Before:
(defun ekg-capture (&optional tags properties subject)
  "Capture a new note, with TAGS and other PROPERTIES.
If SUBJECT is given, force the triple subject to be that value."
  (interactive)
  (let* ((subject (or subject (ekg--generate-id)))
         (buf (get-buffer-create (format "*EKG Capture (note %s)*" subject))))
    (set-buffer buf)
    (funcall ekg-capture-default-mode)
    (ekg-capture-mode 1)
    (setq ekg-note
            (ekg-note-create nil ekg-capture-default-mode
                             (seq-uniq (append
                                        tags
                                        (mapcan (lambda (f) (funcall f)) ekg-capture-auto-tag-funcs)))))
    (setf (ekg-note-id ekg-note) subject)
    (setf (ekg-note-properties ekg-note) properties)
    (ekg-edit-display-metadata)
    (goto-char (point-max))
    (mapc (lambda (tag) (run-hook-with-args 'ekg-note-add-tag-hook tag)) (ekg-note-tags ekg-note))
    (pop-to-buffer buf)))

;; After:
(defun ekg-capture (&optional tags properties subject)
  "Capture a new note, with TAGS and other PROPERTIES.
If SUBJECT is given, force the triple subject to be that value."
  (interactive)
  (let* ((subject (or subject (ekg--generate-id)))
         (buf (get-buffer-create (format "*EKG Capture (note %s)*" subject))))
    (set-buffer buf)
    (funcall ekg-capture-default-mode)
    (ekg-capture-mode 1)
    ;; changes begin here: ↓↓↓
    (setq ekg-note
          (ekg-note-create
           nil nil
           (seq-uniq (append tags (mapcan (lambda (f) (funcall f)) ekg-capture-auto-tag-funcs)))
           properties subject))
    ;; changes end here: ↑↑↑
    (ekg-edit-display-metadata)
    (goto-char (point-max))
    (mapc (lambda (tag) (run-hook-with-args 'ekg-note-add-tag-hook tag)) (ekg-note-tags ekg-note))
    (pop-to-buffer buf)))
```

**Comment by ahyatt** [comment]
The more arguments, the harder to read.  But I agree with the spirit.  What if we turned this into a function that accepted key/val parameters, so you can say `(ekg-capture "My text" :tags ("foo" "bar") :id 123 :mode 'text-mode :title "Test note")`?

**Reply by qingshuizheng** [reply]
I thought of it, but tried to avoid touching the "*-test.el" files, since there are lots of `ekg-note-create` & `ekg-capture` instances  there to take care of.  :-)

I like the idea of using key/val parameters. It is flexible and looks nice and neat.

The only side-effect is that it might break end-users' configs if they've already put some custom defun with `ekg-capture` and `ekg-note-create`.

**Reply by qingshuizheng** [reply]
Just forced push a commit, if you like to see it: https://github.com/qingshuizheng/ekg/commit/05dde84

> (ekg-capture "My text" :tags ("foo" "bar") :id 123 :mode 'text-mode :title "Test note")

1. The syntax is different from what you have mentioned above. If we don't want the quote marks, only defmacro could do that?
```elisp
(ekg-note-create :text "My text 1"
                 :tags '("foo" "bar")
                 :id 123
                 :mode 'text-mode
                 :properties '(:titled/title ("Test 1")))
(ekg-capture :text "My text 2"
             :tags '("ekg" "triples")
             :id 456
             :mode 'org-mode
             :properties '(:titled/title ("Test 2")))
```

2. All `ekg-test-*` tests passed:
```
Selector: "ekg-test-.*"
Passed:  20
Failed:  0
Skipped: 0
Total:   20/20

Started at:   2023-06-09 19:57:53+0800
Finished.
Finished at:  2023-06-09 19:57:56+0800
```

3. Any essential key that is nil, will be converted to the default value:
- `:text` --> ""
- `:tags` --> (ekg-date-tag), ;; even when `ekg-template-tag` set nil
- `:id` --> (ekg--generate-id)
- `:mode` --> ekg-capture-default-mode

4. One niche need to deal:
- When `:text` has text, it completes with the template note text.

**Reply by ahyatt** [reply]
This looks good, please send a pull request!

I don't understand the part about the template note text, but it should be that if you are creating with tags, it should run the `ekg-note-add-tag-hook` after the text is already set in the note, which should then just append to the text.

**Reply by qingshuizheng** [reply]
PR https://github.com/ahyatt/ekg/pull/77 sent.

> which should then just append to the text.

Got you. Additionally, maybe we  need a "\n\n" and the like as the separator between captured text and the text insert by template. And make sure don't insert the separator when capture text is only empty spaces.
